### PR TITLE
PyArrow support

### LINKED ISSUE #375: Support for returning pyarrow objects
It would be great to support returning the following objects to support comprehensive null handling as well as improve throughput, reduce memory usage and increase ecosystem compatibility:

1. `pyarrow.Table` (`query_arrow`?)
2. `pyarrow.RecordBatchReader` (`query_stream`?)

### PR DESCRIPTION
**Pull request by khvn26**
Fixes #375.

In this PR, we add two client methods for querying data as PyArrow objects:

- `query_arrow` returns the whole result as a `pyarrow.Table`.
- `query_arrow_stream` returns a `pyarrow.RecordBatchReader` yielding one record batch per ClickHouse block. Batch size follows the `max_block_size` setting.

Both work on any client. PyArrow is an optional extra installable via `pip install clickhouse-driver[arrow]`.

New APIs accept an optional `arrow_types={'col_name': pa_type}` mapping. If not provided, columns are converted via the built-in type mapping, with the schema built from the query's header block; types outside the mapping fall back to best-effort inference from the first data block:

- Empty results return a typed empty table and the streaming reader exposes `.schema` before any data block is consumed.
- `Nullable(T)` columns come back as proper Arrow nulls and keep their original type — e.g. `Nullable(Int64)` stays a nullable `int64` column, with no dtype degradation as in pandas. With `use_numpy=True`, the wire null mask is packed directly into the Arrow validity bitmap and the values buffer is reused as is.
- With `use_numpy=True`, numeric and datetime64 columns are passed to Arrow without copying. Results are identical with and without numpy.
- `LowCardinality(T)` maps to plain `T`, not `pa.dictionary(...)`, so results are identical whichever code path produced them; dictionary-preserving output could be added later behind an option.
- `JSON` columns have no default representation and require an `arrow_types` entry. `pa.string()` gives JSON text, a struct type gives structured output. JSON text output requires `output_format_native_write_json_as_string=1` or `toJSONString()` in the query. Struct output works with either wire format.

The type mapping table is documented in `docs/features.rst`. Every Arrow field carries its original ClickHouse type as `clickhouse_type` metadata (opt out with `field_metadata=False`).

A partially consumed, or closed, streaming reader doesn't break the client: when the next query starts, the streamed query is cancelled and drained through END_OF_STREAM (the connection is reset if draining fails).

CI gains three `USE_ARROW` jobs: ClickHouse 25.5 / Python 3.12 with numpy, ClickHouse 25.5 / Python 3.13 with pyarrow only (no numpy/pandas), and ClickHouse 20.3 / Python 3.9 with the oldest resolvable pyarrow/numpy pair.

## Microbenchmarks: single-type throughput ceilings

These are synthetic single-type queries over `system.numbers` (1–2 columns, 2–5M rows), timed in-process on a warm client — they show the ceiling each fast path reaches when a query is dominated by that one type.

Local ClickHouse 25.12 in Docker, M-series MacBook, median of 3 runs, rows/s (higher is better):

| shape (rows) | `execute` | `execute` columnar | `query_dataframe` | `query_arrow` | `query_arrow` + numpy | stream + numpy |
|---|---|---|---|---|---|---|
| numeric (5M) | 11.4M | 25.4M | 107M | 24.3M | **117M** | **213M** |
| strings (2M) | 12.8M | 23.1M | 6.8M | 21.0M | **76M** | **86M** |
| nullable (2M) | 10.3M | 15.8M | 25.6M | 13.8M | **232M** | **236M** |
| lowcard (2M) | 14.9M | 32.5M | 53.2M | 27.6M | **59.4M** | 59.5M |
| datetime (2M) | 1.8M | 1.8M | 205M | 1.3M | **174M** | 159M |

For reference, the same queries through clickhouse-connect's `query_arrow` (HTTP, server-side Arrow serialization, `output_format_arrow_string_as_string=1`) on the same machine:

| shape (rows) | this PR | clickhouse-connect |
|---|---|---|
| numeric (5M) | **213M** | 37M |
| strings (2M) | **86M** | 23.3M |
| nullable (2M) | **236M** | 49M |
| lowcard (2M) | **59.4M** | 24M |
| datetime (2M) | **174M** | 86M |

With `use_numpy=True`, `String` columns are read from the wire directly into Arrow offset/data buffers, bypassing Python strings. FixedString, LowCardinality dictionaries, custom encodings and sparse serialization keep the existing path, and Cython-less builds fall back gracefully.

Arrow-specific read paths live in their own column family (`clickhouse_driver/columns/arrow/`), mirroring how `columns/numpy/` relates to the generic columns: the NumPy column classes carry no Arrow logic.

Checklist:

- [x] Add tests that demonstrate the correct behavior of the change. Tests should fail without the change.
- [x] Add or update relevant docs, in the docs folder and in code.
- [x] Ensure PR doesn't contain untouched code reformatting: spaces, etc.
- [x] Run `flake8` and fix issues.
- [x] Run `pytest` no tests failed. See https://clickhouse-driver.readthedocs.io/en/latest/development.html.

### CODE REVIEW

File: tests/arrow/test_types.py [hunk 1]
```diff
@@ -0,0 +1,608 @@
...
+            self.assertEqual(len(cm.output), 1)
+            self.assertIn(
+                'output_format_native_write_json_as_string', cm.output[0]
+            )
+
+    def test_json_as_text_passthrough_does_not_warn(self):
+        records = []
+        handler = logging.Handler()
+        handler.emit = records.append
+        logger = logging.getLogger('clickhouse_driver.arrow.mapping')
```
Comment (xzkostyan) [inline tests/arrow/test_types.py]: `with capture_logging('clickhouse_driver.arrow.mapping', 'INFO'):` may be helpful

File: tests/arrow/test_mapping.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    """
+    Unit tests for reader creation edge cases. No server needed.
+    """
+
+    def setUp(self):
+        if pa is None:
+            self.skipTest('PyArrow package is not installed')
+
+    def make_context(self):
+        from clickhouse_driver.context import Context
```
Comment (xzkostyan) [inline tests/arrow/test_mapping.py]: Can we move local imports to module level?

**Review by xzkostyan — APPROVED** [review]

File: tests/arrow/test_types.py (hunk 1 above)
Comment (khvn26) [reply tests/arrow/test_types.py]: Agreed. fb7744f

File: tests/arrow/test_mapping.py (hunk 2 above)
Comment (khvn26) [reply tests/arrow/test_mapping.py]: Done in 27be89a.

File: tests/arrow/test_types.py (hunk 1 above)
Comment (khvn26) [reply tests/arrow/test_types.py]: Actually removed entirely in bd41e31 — we're raising instead of warning now.
